### PR TITLE
PCHR-2937: Add new (svg) spinner for civihr extensions

### PR DESCRIPTION
**Pull request by bmnepali**
## Overview
This PR add the new spinner (svg) animated with css in  `org.civicrm.bootstrapcivihr` extension so that it can be reused for better consistency across civihr extensions.

## Before
- Previously the gif is shown as spinner.
2937-before-min

## After
- The gif is replaced by the new spinner.
2937-final-min

## Technical Details
1. Add a spinner svg creating a new folder `img/` (`org.civicrm.bootstrapcivihr/img/spinner.svg`).
```html
<svg xmlns="http://www.w3.org/2000/svg" xmlns:xlink="http://www.w3.org/1999/xlink" id="Layer_1" data-name="Layer 1" viewBox="0 0 80 80">
  <path
      id="civiHRSpinner"
      fill="#0071BD"
      d="M40,72C22.4,72,8,57.6,8,40C8,22.4,
       22.4,8,40,8c17.6,0,32,14.4,32,32c0,1.1-0.9,2-2,2
       s-2-0.9-2-2c0-15.4-12.6-28-28-28S12,24.6,12,40s12.6,
       28,28,28c1.1,0,2,0.9,2,2S41.1,72,40,72z">
   </path>
   <title>
      CiviHRSpinner
   </title>
</svg>
```

2. Create a `mixin` for the spinner in `bootstrapcivihr exension` (`org.civicrm.bootstrapcivihr/scss/mixins/_spinner.scss`):
```css
...
@mixin spinner($force-size, $speed: 1.5s, $size: 32px) {
  $size: if($force-size == '', $size, $size !important);
  ...
  height: $size;
  width: $size;
  ...
}
```

3. Override the CiviCRM spinner using the new spinner  in (`org.civicrm.bootstrapcivihr/scss/components/overrides/crm/_spinner.scss`):
```css
/**
 * Overriding the default civicrm core spinner
 * defined in file sites/all/modules/civicrm/css/civicrm.css
 */
.crm-container {
  .dataTables_processing,
  .blockUI.blockOverlay {
    @include spinner('true');
    margin: 11% 50% !important;
  }
}
```

4. Update the loader/spinner directive template in `reqangular extension` (`org.civicrm.reqangular/src/common/templates/loading.html`):
```html
<div>
  <div prevent-animations class="spinner" ng-show="!show"></div>
  <div ng-transclude ng-show="show"></div>
</div>
```
## Comments
As part of this PR, the spinner is applied to all extensions except hrjobcontract and T&A which have their own version of spinner directive. And these two extensions will be using the spinner form bootstrapcivihr extension in respective tasks.

## Tests
✅ Jasmine Tests - passed
✅ JS distributive files - created
✅  CSS distributive files -  created
⏹ Backstop Tests - n/a
⏹ PHP Unit Tests - not needed to run

### CODE REVIEW

File: org.civicrm.bootstrapcivihr/scss/components/_loader.scss [hunk 1]
```diff
@@ -0,0 +1,11 @@
+/**
```
Comment (AkA84) [inline org.civicrm.bootstrapcivihr/scss/components/_loader.scss]: this is not a component, and even if it was the name of the file would not reflect the name of the component

You are overriding a core screen style, create a `scss/overrides/crm` folder and place it in a aptly-named partial

File: org.civicrm.bootstrapcivihr/scss/components/_spinner.scss [hunk 2]
```diff
@@ -0,0 +1,3 @@
+.spinner {
+  @include spinner(1s);
```
Comment (AkA84) [inline org.civicrm.bootstrapcivihr/scss/components/_spinner.scss]: if it's always `1s` both here and in the override, then just put it as a default value in the mixin

File: org.civicrm.bootstrapcivihr/scss/mixins/_spinner.scss [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+ * $height Width of the spinner
+ */
+@mixin spinner($speed: $speed, $height: $default-height, $width: $default-width) {
+  animation: spin $speed linear infinite;
+  background: url(../img/spinner.svg) no-repeat center center !important;
+  display: block;
+  margin: auto;
+
+  @if $height {
+    height: $height !important;
```
Comment (AkA84) [inline org.civicrm.bootstrapcivihr/scss/mixins/_spinner.scss]: use a ternary operator (applies for all the other instances)
```scss
$height: if($height, /*value*/, /*other value*/);
```

File: org.civicrm.bootstrapcivihr/scss/mixins/_spinner.scss [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+  } @else {
```
Comment (AkA84) [inline org.civicrm.bootstrapcivihr/scss/mixins/_spinner.scss]: actually i don't get this if/else at all, you are defining already the default values, so what's the point of this?

File: org.civicrm.bootstrapcivihr/scss/mixins/_spinner.scss [hunk 5]
```diff
@@ -0,0 +1,39 @@
+// Spinner related variables
+$default-height: 32px;
+$default-width: 32px;
+$speed: 1s;
```
Comment (AkA84) [inline org.civicrm.bootstrapcivihr/scss/mixins/_spinner.scss]: just define these are the default values, no need for these variables

File: org.civicrm.bootstrapcivihr/scss/mixins/_spinner.scss [hunk 6]
```diff
@@ -0,0 +1,39 @@
+// Spinner related variables
+$default-height: 32px;
+$default-width: 32px;
```
Comment (AkA84) [inline org.civicrm.bootstrapcivihr/scss/mixins/_spinner.scss]: why would the width be ever different than the height? the spinner should always be a square

File: org.civicrm.bootstrapcivihr/scss/mixins/_spinner.scss (hunk 4 above)
Comment (bmnepali) [reply org.civicrm.bootstrapcivihr/scss/mixins/_spinner.scss]: Right! thanks for catching it! in fact we won't be needing it!

File: org.civicrm.bootstrapcivihr/scss/civihr.scss [hunk 7]
```diff
@@ -6,6 +6,8 @@
 @import "fonts/*";
 
 @import "overrides/style/*";
+@import "mixins/*";
```
Comment (AkA84) [inline org.civicrm.bootstrapcivihr/scss/civihr.scss]:
```scss
@import "fonts/*";
@import "mixins/*";

@import "overrides/crm/*";
@import "overrides/style/*";

File: org.civicrm.bootstrapcivihr/scss/components/_spinner.scss [hunk 8]
```diff
@@ -0,0 +1,3 @@
+.spinner {
+  @include spinner(1.5s);
```
Comment (AkA84) [inline org.civicrm.bootstrapcivihr/scss/components/_spinner.scss]: now why did you change it to 1.5s, when before it was 1s?

if *both* implementations of the spinner requires 1.5s, then just define that as the default value and don't pass anything to the function

File: org.civicrm.bootstrapcivihr/scss/mixins/_spinner.scss [hunk 9]
```diff
@@ -0,0 +1,24 @@
...
+ * Spinner mixin
+ * $speed Speed to rorate the spinner in seconds
+ * $height Height of the spinner
+ * $height Width of the spinner
+ */
+@mixin spinner($speed: 1s, $height: 32px, $width: 32px) {
+  animation: spin $speed linear infinite;
+  background: url(../img/spinner.svg) no-repeat center center !important;
+  display: block;
+  height: $height !important;
```
Comment (AkA84) [inline org.civicrm.bootstrapcivihr/scss/mixins/_spinner.scss]: why `!important`?

File: org.civicrm.bootstrapcivihr/scss/mixins/_spinner.scss [hunk 10]
```diff
@@ -0,0 +1,24 @@
+/**
+ * Spinner mixin
+ * $speed Speed to rorate the spinner in seconds
+ * $height Height of the spinner
+ * $height Width of the spinner
+ */
+@mixin spinner($speed: 1s, $height: 32px, $width: 32px) {
```
Comment (AkA84) [inline org.civicrm.bootstrapcivihr/scss/mixins/_spinner.scss]: you are still able to define different width and height, which is something we *do not* want to happen.

File: org.civicrm.bootstrapcivihr/scss/components/_spinner.scss (hunk 8 above)
Comment (bmnepali) [reply org.civicrm.bootstrapcivihr/scss/components/_spinner.scss]: > now why did you change it to 1.5s, when before it was 1s?
Previously, the speed was `1s` but i realized the speed should  actually be `1.5s` as per @jon's comment. so i changed it to later.

File: org.civicrm.bootstrapcivihr/scss/mixins/_spinner.scss (hunk 10 above)
Comment (bmnepali) [reply org.civicrm.bootstrapcivihr/scss/mixins/_spinner.scss]: Correct! lets pass a param called `$size: 32px`  whose default size will be `32px`.

File: org.civicrm.bootstrapcivihr/scss/mixins/_spinner.scss (hunk 9 above)
Comment (bmnepali) [reply org.civicrm.bootstrapcivihr/scss/mixins/_spinner.scss]: The core CiviCRM spinner already has defined height and width to be `100%`. So, to override it `!important` is used here.

File: org.civicrm.bootstrapcivihr/scss/civihr.scss [hunk 11]
```diff
@@ -4,8 +4,11 @@
 @import "SCSSROOT/org.civicrm.shoreditch/base/scss/vendor/bootstrap/mixins";
 
 @import "fonts/*";
+@import "mixins/*";
 
+@import "overrides/crm/*";
 @import "overrides/style/*";
+
```
Comment (AkA84) [inline org.civicrm.bootstrapcivihr/scss/civihr.scss]: don't leave this empty line

File: org.civicrm.bootstrapcivihr/scss/mixins/_spinner.scss (hunk 9 above)
Comment (AkA84) [reply org.civicrm.bootstrapcivihr/scss/mixins/_spinner.scss]: why can't you override it by just using a "heavier" selector? It's not good to have "!important" in a generic mixin just to fix a very specific case

Comment (bmnepali) [reply org.civicrm.bootstrapcivihr/scss/mixins/_spinner.scss]: > why can't you override it by just using a "heavier" selector?

Since the styles are added inline and there is this only one element and has no nested elements, don't think we can target styles using any selectors leaving behind the only option to use `!important`.
screen shot 2017-11-20 at 8 37 30 pm

However, I have refactored the mixin avoiding the use of `!important`. And overriding the size of core civicrm spinner  in specific overriding styles.

File: org.civicrm.bootstrapcivihr/scss/overrides/_variables.scss [hunk 12]
```diff
@@ -14,3 +14,4 @@ $steps-height: 24px;
 $steps-compact-height: 28px;
 
 $steps-gutter: ($steps-height / 4);
+$spinner: ($steps-height / 4);
```
Comment (AkA84) [inline org.civicrm.bootstrapcivihr/scss/overrides/_variables.scss]: what is this for, you don't seem to be using it anywhere else and it wasn't here in the previous commits

File: org.civicrm.bootstrapcivihr/scss/overrides/crm/_spinner.scss [hunk 13]
```diff
@@ -0,0 +1,13 @@
+/**
+ * Overriding the default civicrm core spinner
+ * defined in file sites/all/modules/civicrm/css/civicrm.css
+ */
+.crm-container {
+  .dataTables_processing,
+  .blockUI.blockOverlay {
+    @include spinner();
+    height: 32px !important;
```
Comment (AkA84) [inline org.civicrm.bootstrapcivihr/scss/overrides/crm/_spinner.scss]: that's not what i was talking about, and it's wrong cause what's the point of providing the `$size` property in the mixin if you then redefine width and height outside of it. Use a more specific/heavier selector so that you don't need to use `!important` in the css

File: org.civicrm.bootstrapcivihr/scss/overrides/_variables.scss (hunk 12 above)
Comment (bmnepali) [reply org.civicrm.bootstrapcivihr/scss/overrides/_variables.scss]: I am not sure why its changed!  i will fix it. I wont be needing it.
